nupic/research/frameworks/greedy_infomax/models/BilinearInfo.py

```diff
@@ -0,0 +1,198 @@
+                          Skipping 1 step prevents overlapping samples from being
+                          compared with the default overlap. This parameter should
+                          probably be increased for larger overlap values.
+        """
+        batch_size = z.shape[0]
+
+        # For each k in k_predictions, store a set of log_fk and true_f values
+        log_f_list, true_f_list = [], []
+        for k in range(1, self.k_predictions + 1):
+            # Compute log f(c_t, x_{t+k}) = z^T_{t+k} W_k c_t
+
+            # First half of bilinear model, compute z^T_{t+k} W_k:
+            ztwk = (
+                self.W_k[k - 1]  # 1x1 convolution
+                .forward(z[:, :, (k + skip_step) :, :])  # Bx, C , H , W
+                .permute(2, 3, 0, 1)  # H, W, Bx, C
+                .contiguous()
+            )  # y, x, b, c
+
+            # Take random samples from z_{t+k} W_k as contrastive samples
+            ztwk_shuf = ztwk.view(
+                ztwk.shape[0] * ztwk.shape[1] * ztwk.shape[2], ztwk.shape[3]
+            )  # y * x * batch, c
+            rand_index = torch.randint(
+                ztwk_shuf.shape[0],  # y *  x * batch
+                (ztwk_shuf.shape[0] * self.negative_samples, 1),
+                dtype=torch.long,
+                device=ztwk_shuf.device,
+            )
+            rand_index = rand_index.repeat(1, ztwk_shuf.shape[1])
+            ztwk_shuf = torch.gather(
+                ztwk_shuf, dim=0, index=rand_index, out=None
+            )  # y * x * b * n, c
+            ztwk_shuf = ztwk_shuf.view(
+                ztwk.shape[0],
+                ztwk.shape[1],
+                ztwk.shape[2],
+                self.negative_samples,
+                ztwk.shape[3],
+            ).permute(
+                0, 1, 2, 4, 3
+            )  # y, x, b, c, n
+
+            # Multiply ztwk and context for full bilinear model
+            context = (
+                c[:, :, : -(k + skip_step), :].permute(2, 3, 0, 1).unsqueeze(-2)
+            )  # y, x, b, 1, c
+
+            log_fk_main = torch.matmul(context, ztwk.unsqueeze(-1)).squeeze(
+                -2
+            )  # y, x, b, 1
+
+            log_fk_shuf = torch.matmul(context, ztwk_shuf).squeeze(-2)  # y, x, b, n
+
+            # Stack negative samples below positive samples
+            log_fk = torch.cat((log_fk_main, log_fk_shuf), 3)  # y, x, b, 1+n
+            log_fk = log_fk.permute(2, 3, 0, 1)  # b, 1+n, y, x
+
+            # Positive samples are at index 0
+            true_fk = torch.zeros(
+                (batch_size, log_fk.shape[-2], log_fk.shape[-1]),
+                dtype=torch.long,
+                device=log_fk.device,
+            )  # b, y, x
+
+            # #append results to list
+            log_f_list.append(log_fk)
+            true_f_list.append(true_fk)
+
+        return log_f_list, true_f_list
+
+
+class SparseBilinearInfo(BilinearInfo):
+    def __init__(
+        self,
+        in_channels,
```

Comment:
Please add documentation for this class, especially since since the `__init__` here has has args like `sparsity` that aren't used in the parent class.

Reply:
Done